cedar-lean/Cedar/Partial/Evaluator.lean

```diff
@@ -189,16 +189,99 @@ def evaluate (x : Partial.Expr) (req : Partial.Request) (es : Partial.Entities)
     let pval₁ ← Partial.evaluate x₁ req es
     evaluateGetAttr pval₁ a es
   | .set xs          => do
-    let vs ← xs.mapM₁ (fun ⟨x₁, _⟩ => Partial.evaluate x₁ req es)
-    match vs.mapM (fun pval => match pval with | .value v => some v | .residual _ => none) with
-    | some vs => .ok (.value (Set.make vs))
-    | none    => .ok (.residual (Partial.Expr.set (vs.map Partial.Value.asPartialExpr)))
+    let pvs ← xs.mapM₁ (λ ⟨x₁, _⟩ => Partial.evaluate x₁ req es)
+    match pvs.mapM (λ pval => match pval with | .value v => some v | .residual _ => none) with
+    | some vs => .ok $ .value $ Set.make vs
+    | none    => .ok $ .residual $ .set pvs
   | .record axs      => do
-    let avs ← axs.mapM₂ (fun ⟨(a₁, x₁), _⟩ => Partial.bindAttr a₁ (Partial.evaluate x₁ req es))
-    match avs.mapM (fun (a, pval) => match pval with | .value v => some (a, v) | .residual _ => none) with
-    | some avs => .ok (.value (Map.make avs))
-    | none     => .ok (.residual (Partial.Expr.record (avs.map fun (a, v) => (a, v.asPartialExpr))))
+    let apvs ← axs.mapM₂ (λ ⟨(a₁, x₁), _⟩ => Partial.bindAttr a₁ (Partial.evaluate x₁ req es))
+    match apvs.mapM (λ (a, pval) => match pval with | .value v => some (a, v) | .residual _ => none) with
+    | some avs => .ok $ .value $ Map.make avs
+    | none     => .ok $ .residual $ .record apvs
   | .call xfn xs     => do
-    let pvs ← xs.mapM₁ (fun ⟨x₁, _⟩ => Partial.evaluate x₁ req es)
+    let pvs ← xs.mapM₁ (λ ⟨x₁, _⟩ => Partial.evaluate x₁ req es)
     evaluateCall xfn pvs
-  | .unknown u       => .ok (.residual (Partial.Expr.unknown u))
+
+mutual
+
+/--
+  Evaluate a `Partial.Value`, possibly reducing it. For instance, `3 + 5` will
+  evaluate to `8`. This can be relevant if a substitution was recently made on
+  the `Partial.Value`.
+-/
+def evaluateValue (pv : Partial.Value) (es : Partial.Entities) : Result Partial.Value :=
```

Comment:
What's this referring to -- unifying `evaluate` and `evaluateValue`?  This PR proposes separating them, so that when `evaluateGetAttr` needs to evaluate any residuals it gets from entity attributes, that is not itself a new mutually recursive call all the way back to `evaluate`.  (Also, so that `evaluateValue` doesn't need the `request` argument.)

Reply:
It was referring to the general question (answered above, thanks!) about unifying `Partial.Value` and `ResidualExpr` and consequently some of the code that works on them.